In [31]:
%pip install -q kagglehub
import kagglehub
import pandas as pd
import os

path = kagglehub.dataset_download(
    "olistbr/brazilian-ecommerce"
)

print(path)


[notice] A new release of pip is available: 25.2 -> 26.2.1
[notice] To update, run: python.exe -m pip install --upgrade pip


Note: you may need to restart the kernel to use updated packages.
C:\Users\Usuario\.cache\kagglehub\datasets\olistbr\brazilian-ecommerce\versions\2


In [32]:
files = os.listdir(path)

for file in files:
    print(file)

olist_customers_dataset.csv
olist_geolocation_dataset.csv
olist_orders_dataset.csv
olist_order_items_dataset.csv
olist_order_payments_dataset.csv
olist_order_reviews_dataset.csv
olist_products_dataset.csv
olist_sellers_dataset.csv
product_category_name_translation.csv


In [33]:
#asi se cargan las tablas en un diccionario, donde la clave es el nombre de la tabla y el valor es el DataFrame

tables = {}
for file in files:
    if file.endswith(".csv"):
        table_name = file.replace(".csv", "")
        table_path = os.path.join(path, file)
        tables[table_name] = pd.read_csv(table_path)
        print(f"Loaded table: {table_name}")

tables["olist_customers_dataset"].head()

Loaded table: olist_customers_dataset
Loaded table: olist_geolocation_dataset
Loaded table: olist_orders_dataset
Loaded table: olist_order_items_dataset
Loaded table: olist_order_payments_dataset
Loaded table: olist_order_reviews_dataset
Loaded table: olist_products_dataset
Loaded table: olist_sellers_dataset
Loaded table: product_category_name_translation


,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
0,06b8999e2fba1a1fbc88172c00ba8bc7,861eff4711a542e4b93843c6dd7febb0,14409,franca,SP
1,18955e83d337fd6b2def6b18a428ac77,290c77bc529b7ac935b93aa66c333dc3,9790,sao bernardo do campo,SP
2,4e7b3e00288586ebd08712fdd0374a03,060e732b5b29e8181a18229c7b0b2b5e,1151,sao paulo,SP
3,b2b6027bc5c5109e529d4dc6358b12c3,259dac757896d24d7702b9acbbff3f3c,8775,mogi das cruzes,SP
4,4f2d8ab171c80ec8364f7c12e35b23ad,345ecd01c38d18a9036ed96c73b8d066,13056,campinas,SP


In [34]:
#ahora vamos a ver el tamanio de las tablas
for name, df in tables.items():
    print(f"Table: {name}, Rows: {df.shape[0]}, Columns: {df.shape[1]}")

Table: olist_customers_dataset, Rows: 99441, Columns: 5
Table: olist_geolocation_dataset, Rows: 1000163, Columns: 5
Table: olist_orders_dataset, Rows: 99441, Columns: 8
Table: olist_order_items_dataset, Rows: 112650, Columns: 7
Table: olist_order_payments_dataset, Rows: 103886, Columns: 5
Table: olist_order_reviews_dataset, Rows: 99224, Columns: 7
Table: olist_products_dataset, Rows: 32951, Columns: 9
Table: olist_sellers_dataset, Rows: 3095, Columns: 4
Table: product_category_name_translation, Rows: 71, Columns: 2


In [35]:
#primer resumen basico automatico, seria un data quality overview, que nos da una idea de la calidad de los datos y de la estructura de las tablas
summary = {}

for name, df in tables.items():
    summary[name] = {
        "rows": df.shape[0],
        "columns": df.shape[1],
        "missing_values": df.isnull().sum().sum(),
        "duplicate_rows": df.duplicated().sum(),
        "data_types": df.dtypes.value_counts().to_dict()
    }
summary_df = pd.DataFrame.from_dict(summary, orient='index')
summary_df

,rows,columns,missing_values,duplicate_rows,data_types
olist_customers_dataset,99441,5,0,0,"{str: 4, int64: 1}"
olist_geolocation_dataset,1000163,5,0,261831,"{float64: 2, str: 2, int64: 1}"
olist_orders_dataset,99441,8,4908,0,{str: 8}
olist_order_items_dataset,112650,7,0,0,"{str: 4, float64: 2, int64: 1}"
olist_order_payments_dataset,103886,5,0,0,"{str: 2, int64: 2, float64: 1}"
olist_order_reviews_dataset,99224,7,145903,0,"{str: 6, int64: 1}"
olist_products_dataset,32951,9,2448,0,"{float64: 7, str: 2}"
olist_sellers_dataset,3095,4,0,0,"{str: 3, int64: 1}"
product_category_name_translation,71,2,0,0,{str: 2}


In [36]:
orders = tables["olist_orders_dataset"].copy()
orders.info()
orders.head()

<class 'pandas.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype
---  ------                         --------------  -----
 0   order_id                       99441 non-null  str  
 1   customer_id                    99441 non-null  str  
 2   order_status                   99441 non-null  str  
 3   order_purchase_timestamp       99441 non-null  str  
 4   order_approved_at              99281 non-null  str  
 5   order_delivered_carrier_date   97658 non-null  str  
 6   order_delivered_customer_date  96476 non-null  str  
 7   order_estimated_delivery_date  99441 non-null  str  
dtypes: str(8)
memory usage: 21.9 MB


,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18 00:00:00
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13 00:00:00
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04 00:00:00
3,949d5b44dbf5de918fe9c16f97b45f8a,f88197465ea7920adcdbec7375364d82,delivered,2017-11-18 19:28:06,2017-11-18 19:45:59,2017-11-22 13:39:59,2017-12-02 00:28:42,2017-12-15 00:00:00
4,ad21c59c0840e6cb83a9ceb5573f8159,8ab97904e6daea8866dbdbc4fb7aad2c,delivered,2018-02-13 21:18:39,2018-02-13 22:20:29,2018-02-14 19:46:34,2018-02-16 18:17:02,2018-02-26 00:00:00


In [58]:
nulls = pd.DataFrame({
    "columna": orders.columns,
    "cantidad_nulos": orders.isnull().sum(),
    "porcentaje_nulos": orders.isnull().mean() * 100
})

nulls

# condicion = orders["order_approved_at"].isnull() & orders["order_delivered_carrier_date"].notnull()
# orders[condicion][["order_id", "order_approved_at", "order_delivered_carrier_date", "order_purchase_timestamp"]]

# condicion = orders["order_approved_at"].isnull()
# approved_nulls = orders[condicion]
# approved_nulls.head()

# print(f"Cantidad de filas con order_approved_at nulo: {approved_nulls.shape[0]}")
# approved_nulls["order_status"].value_counts()

# condicion2 = approved_nulls["order_status"] == "delivered"
# delivered_nulls = approved_nulls[condicion2]
# print(f"Cantidad de filas con order_approved_at nulo y order_status 'delivered': {delivered_nulls.shape[0]}")
# delivered_nulls[["order_id", "order_approved_at", "order_delivered_carrier_date", "order_purchase_timestamp"]]


,columna,cantidad_nulos,porcentaje_nulos
order_id,order_id,0,0.000000
customer_id,customer_id,0,0.000000
order_status,order_status,0,0.000000
order_purchase_timestamp,order_purchase_timestamp,0,0.000000
order_approved_at,order_approved_at,160,0.160899
order_delivered_carrier_date,order_delivered_carrier_date,1783,1.793023
order_delivered_customer_date,order_delivered_customer_date,2965,2.981668
order_estimated_delivery_date,order_estimated_delivery_date,0,0.000000


In [52]:
condicion_created = (
    approved_nulls["order_status"] == "created"
)
created_nulls = approved_nulls[condicion_created]
print(f"Cantidad de filas con order_approved_at nulo y order_status 'created': {created_nulls.shape[0]}")
created_nulls[
    [
        "order_id",
        "order_status",
        "order_purchase_timestamp",
        "order_approved_at",
        "order_delivered_carrier_date",
        "order_delivered_customer_date",
        "order_estimated_delivery_date"
    ]
]

Cantidad de filas con order_approved_at nulo y order_status 'created': 5


,order_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
7434,b5359909123fa03c50bdb0cfed07f098,created,2017-12-05 01:07:52,NaN,NaN,NaN,2018-01-11 00:00:00
9238,dba5062fbda3af4fb6c33b1e040ca38f,created,2018-02-09 17:21:04,NaN,NaN,NaN,2018-03-07 00:00:00
21441,7a4df5d8cff4090e541401a20a22bb80,created,2017-11-25 11:10:33,NaN,NaN,NaN,2017-12-12 00:00:00
55086,35de4050331c6c644cddc86f4f2d0d64,created,2017-12-05 01:07:58,NaN,NaN,NaN,2018-01-08 00:00:00
58958,90ab3e7d52544ec7bc3363c82689965f,created,2017-11-06 13:12:34,NaN,NaN,NaN,2017-12-01 00:00:00


In [55]:
delivered_nulls[
    ["order_purchase_timestamp",
    "order_approved_at",
    "order_delivered_carrier_date",
    "order_delivered_customer_date",
    "order_estimated_delivery_date"]
]

,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
5323,2017-02-18 14:40:00,NaN,2017-02-23 12:04:47,2017-03-01 13:25:33,2017-03-17 00:00:00
16567,2017-02-18 12:45:31,NaN,2017-02-23 09:01:52,2017-03-02 10:05:06,2017-03-21 00:00:00
19031,2017-02-18 13:29:47,NaN,2017-02-22 16:25:25,2017-03-01 08:07:38,2017-03-17 00:00:00
22663,2017-02-18 16:48:35,NaN,2017-02-22 11:23:10,2017-03-09 07:28:47,2017-03-31 00:00:00
23156,2017-02-17 13:05:55,NaN,2017-02-22 11:23:11,2017-03-02 11:09:19,2017-03-20 00:00:00
26800,2017-01-19 12:48:08,NaN,2017-01-25 14:56:50,2017-01-30 18:16:01,2017-03-01 00:00:00
38290,2017-02-19 01:28:47,NaN,2017-02-23 03:11:48,2017-03-02 03:41:58,2017-03-27 00:00:00
39334,2017-02-18 11:04:19,NaN,2017-02-23 07:23:36,2017-03-02 16:15:23,2017-03-22 00:00:00
48401,2017-01-19 22:26:59,NaN,2017-01-27 11:08:05,2017-02-06 14:22:19,2017-03-16 00:00:00
61743,2017-02-17 17:21:55,NaN,2017-02-22 11:42:51,2017-03-03 12:16:03,2017-03-20 00:00:00


Auditoria de missing values:
De los 160 pedidos sin fecha de aprobación, 141 están cancelados y 5 permanecen en estado created, por lo que la ausencia puede ser coherente con su flujo. Sin embargo, existen 14 pedidos con estado delivered que sí poseen fechas de despacho y entrega, pero carecen de order_approved_at, lo que indica una inconsistencia o pérdida de información en ese atributo.

In [66]:
carrier_nulls = orders[orders["order_delivered_carrier_date"].isna()]
carrier_nulls["order_status"].value_counts()

carrier_nulls[carrier_nulls["order_status"] == "delivered"]

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
73222,2aa91108853cecb43c84a5dc5b277475,afeb16c7f46396c0ed54acb45ccaaa40,delivered,2017-09-29 08:52:58,2017-09-29 09:07:16,NaN,2017-11-20 19:44:47,2017-11-14 00:00:00
92643,2d858f451373b04fb5c984a1cc2defaf,e08caf668d499a6d643dafd7c5cc498a,delivered,2017-05-25 23:22:43,2017-05-25 23:30:16,NaN,NaN,2017-06-23 00:00:00


Los valores nulos de order_delivered_carrier_date son, en su mayoría, consistentes con la lógica del proceso de Olist, ya que corresponden a pedidos que fueron cancelados, no estuvieron disponibles o todavía no alcanzaron la etapa de envío. Sin embargo, se detectaron dos pedidos con estado delivered y sin fecha de entrega al transportista. Uno posee fecha de entrega al cliente, lo que constituye una inconsistencia evidente; el otro tampoco posee fecha de entrega al cliente, por lo que no es posible determinar si el error se encuentra en el estado del pedido o en las fechas registradas. Debido a la falta de evidencia para reconstruir estas fechas, se mantienen los valores nulos y se consideran anomalías de calidad de datos.


In [69]:
delivered_customer_nulls = orders[orders["order_delivered_customer_date"].isna()]
delivered_customer_nulls["order_status"].value_counts()
delivered_customer_nulls[delivered_customer_nulls["order_status"] == "delivered"]

delivered_customer_anomalies = orders[
    (orders["order_status"] == "delivered") &
    (orders["order_delivered_customer_date"].isna())
]

delivered_customer_anomalies


,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
3002,2d1e2d5bf4dc7227b3bfebb81328c15f,ec05a6d8558c6455f0cbbd8a420ad34f,delivered,2017-11-28 17:44:07,2017-11-28 17:56:40,2017-11-30 18:12:23,NaN,2017-12-18 00:00:00
20618,f5dd62b788049ad9fc0526e3ad11a097,5e89028e024b381dc84a13a3570decb4,delivered,2018-06-20 06:58:43,2018-06-20 07:19:05,2018-06-25 08:05:00,NaN,2018-07-16 00:00:00
43834,2ebdfc4f15f23b91474edf87475f108e,29f0540231702fda0cfdee0a310f11aa,delivered,2018-07-01 17:05:11,2018-07-01 17:15:12,2018-07-03 13:57:00,NaN,2018-07-30 00:00:00
79263,e69f75a717d64fc5ecdfae42b2e8e086,cfda40ca8dd0a5d486a9635b611b398a,delivered,2018-07-01 22:05:55,2018-07-01 22:15:14,2018-07-03 13:57:00,NaN,2018-07-30 00:00:00
82868,0d3268bad9b086af767785e3f0fc0133,4f1d63d35fb7c8999853b2699f5c7649,delivered,2018-07-01 21:14:02,2018-07-01 21:29:54,2018-07-03 09:28:00,NaN,2018-07-24 00:00:00
92643,2d858f451373b04fb5c984a1cc2defaf,e08caf668d499a6d643dafd7c5cc498a,delivered,2017-05-25 23:22:43,2017-05-25 23:30:16,NaN,NaN,2017-06-23 00:00:00
97647,ab7c89dc1bf4a1ead9d6ec1ec8968a84,dd1b84a7286eb4524d52af4256c0ba24,delivered,2018-06-08 12:09:39,2018-06-08 12:36:39,2018-06-12 14:10:00,NaN,2018-06-26 00:00:00
98038,20edc82cf5400ce95e1afacc25798b31,28c37425f1127d887d7337f284080a0f,delivered,2018-06-27 16:09:12,2018-06-27 16:29:30,2018-07-03 19:26:00,NaN,2018-07-19 00:00:00


Se analizaron los valores nulos de order_delivered_customer_date. La mayoría son coherentes con la lógica del proceso de negocio, ya que corresponden a pedidos que todavía no alcanzaron la etapa de entrega (created, approved, invoiced, processing y shipped) o cuyo proceso fue interrumpido (canceled y unavailable). Sin embargo, se detectaron 8 pedidos con estado delivered que no poseen fecha de entrega al cliente. En 7 de ellos existe una fecha de entrega al transportista, mientras que en 1 tampoco se registra dicha fecha. Estos registros se consideran anomalías de calidad de datos. Debido a que no existe evidencia suficiente para reconstruir la fecha real de entrega, se conservan los valores nulos y no se utiliza order_estimated_delivery_date como sustituto.

In [72]:
date_columns = [
    "order_purchase_timestamp",
    "order_approved_at",
    "order_delivered_carrier_date",
    "order_delivered_customer_date",
    "order_estimated_delivery_date"
]

for col in date_columns:
    orders[col] = pd.to_datetime(orders[col], errors='coerce')

orders[date_columns].dtypes

order_purchase_timestamp         datetime64[us]
order_approved_at                datetime64[us]
order_delivered_carrier_date     datetime64[us]
order_delivered_customer_date    datetime64[us]
order_estimated_delivery_date    datetime64[us]
dtype: object

Conversión de columnas temporales a tipo datetime
Las columnas relacionadas con fechas y horas se encontraban originalmente almacenadas como tipo str. Para poder realizar operaciones temporales, comparaciones cronológicas y cálculos de duración, se convirtieron al tipo datetime utilizando pd.to_datetime().

Se transformaron las siguientes columnas: order_purchase_timestamp, order_approved_at, order_delivered_carrier_date, order_delivered_customer_date y order_estimated_delivery_date.

Se utilizó el parámetro errors="coerce" para que cualquier valor no interpretable como fecha sea convertido en NaT, preservando así los valores faltantes dentro de un tipo de dato temporal adecuado. Luego de la conversión, se verificó que las cinco columnas quedaran almacenadas como datetime64[us].

Esta transformación permite continuar con controles de calidad más avanzados, como verificar el orden lógico de los eventos del pedido y calcular tiempos transcurridos entre las distintas etapas del proceso.

In [103]:
orders["order_status"].value_counts()

# verificacion de la coherencia cronologica de las fechas, por ejemplo, que la fecha de aprobacion no sea posterior a la fecha de entrega al cliente
# purchase <= approved <= carrier <= customer
#regla 1 purchase <= approved


approved_before_purchase = orders[
    (orders["order_approved_at"] < orders["order_purchase_timestamp"])
]
approved_before_purchase.shape[0]
#el resultado es 0, lo que significa que no hay registros donde la fecha de aprobación sea anterior a la fecha de compra, lo cual es coherente.

#regla 2 approved <= carrier
carrier_before_approved = orders[
    (orders["order_delivered_carrier_date"] < orders["order_approved_at"])
]
carrier_before_approved.shape[0]
#el resultado es 1359, lo que significa que hay 1359 registros donde la fecha de entrega por el transportista sea anterior a la fecha de aprobación, lo cual es incoherente.

#regla 3 carrier <= customer
customer_before_carrier = orders[
    (orders["order_delivered_customer_date"] < orders["order_delivered_carrier_date"])
]
customer_before_carrier.shape[0]
#el resultado es 23, lo que significa que hay 23 registros donde la fecha de entrega al cliente sea anterior a la fecha de entrega por el transportista, lo cual es incoherente.


carrier_before_approved = orders[
    orders["order_delivered_carrier_date"] < orders["order_approved_at"]
].copy()

carrier_before_approved["time_difference"] = (
    carrier_before_approved["order_approved_at"]
    - carrier_before_approved["order_delivered_carrier_date"]
)
carrier_before_approved["time_difference"].describe()

count                      1359
mean     1 days 00:45:07.153053
std      4 days 19:18:28.055754
min             0 days 00:00:21
25%      0 days 01:24:55.500000
50%             0 days 17:10:04
75%             1 days 01:57:24
max           171 days 05:15:22
Name: time_difference, dtype: object


Se verificó la coherencia cronológica entre las principales fechas del proceso de los pedidos, considerando como secuencia esperada purchase <= approved <= carrier <= customer. No se encontraron pedidos cuya fecha de aprobación fuera anterior a la fecha de compra. Sin embargo, se detectaron 1359 registros donde la fecha de entrega al transportista es anterior a la aprobación del pedido y 23 registros donde la fecha de entrega al cliente es anterior a la fecha registrada de entrega al transportista. Estos casos se consideran anomalías cronológicas potenciales y requieren un análisis adicional de las diferencias temporales antes de clasificarlos definitivamente como errores de calidad de datos.

Se identificaron 1359 pedidos cuya fecha de entrega al transportista es anterior a la fecha de aprobación. Las diferencias presentan una alta dispersión: van desde 21 segundos hasta aproximadamente 171 días. La mediana es de aproximadamente 17 horas y el 75 % de los casos presenta una diferencia inferior a aproximadamente 1 día y 2 horas. La existencia de valores extremos considerablemente mayores sugiere que las anomalías podrían tener distintas causas, por lo que es necesario analizar la distribución de las diferencias antes de determinar su tratamiento.